tf/edgeml/trainer/bonsaiTrainer.py

```diff
@@ -0,0 +1,352 @@
+                    self.__Zth: self.__thrsdZ, self.__Tth: self.__thrsdT}
+        sess.run(self.hardThresholdGroup, feed_dict=fd_thrsd)
+
+    def runSparseTraining(self, sess):
+        '''
+        Function to run the Sparse Retraining routine on Bonsai Obj
+        '''
+        currW = self.bonsaiObj.W.eval()
+        currV = self.bonsaiObj.V.eval()
+        currZ = self.bonsaiObj.Z.eval()
+        currT = self.bonsaiObj.T.eval()
+
+        newW = utils.copySupport(self.__thrsdW, currW)
+        newV = utils.copySupport(self.__thrsdV, currV)
+        newZ = utils.copySupport(self.__thrsdZ, currZ)
+        newT = utils.copySupport(self.__thrsdT, currT)
+
+        fd_st = {self.__Wth: newW, self.__Vth: newV,
+                 self.__Zth: newZ, self.__Tth: newT}
+        sess.run(self.sparseRetrainGroup, feed_dict=fd_st)
+
+    def assertInit(self):
+        err = "sparsity must be between 0 and 1"
+        assert self.sW >= 0 and self.sW <= 1, "W " + err
+        assert self.sV >= 0 and self.sV <= 1, "V " + err
+        assert self.sZ >= 0 and self.sZ <= 1, "Z " + err
+        assert self.sT >= 0 and self.sT <= 1, "T " + err
+        errMsg = "Dimension Mismatch, Y has to be [_, " + \
+            str(self.bonsaiObj.numClasses) + "]"
+        errCont = " numClasses are 1 in case of Binary case by design"
+        assert (len(self.Y.shape) == 2 and
+                self.Y.shape[1] == self.bonsaiObj.numClasses), errMsg + errCont
+
+    # Function to get aimed model size
+    def getModelSize(self):
+        nnzZ, sizeZ, sparseZ = utils.countnnZ(self.bonsaiObj.Z, self.sZ)
+        nnzW, sizeW, sparseW = utils.countnnZ(self.bonsaiObj.W, self.sW)
+        nnzV, sizeV, sparseV = utils.countnnZ(self.bonsaiObj.V, self.sV)
+        nnzT, sizeT, sparseT = utils.countnnZ(self.bonsaiObj.T, self.sT)
+
+        totalnnZ = (nnzZ + nnzT + nnzV + nnzW)
+        totalSize = (sizeZ + sizeW + sizeV + sizeT)
+        hasSparse = (sparseW or sparseV or sparseT or sparseZ)
+        return totalnnZ, totalSize, hasSparse
+
+    def train(self, batchSize, totalEpochs, sess,
+              Xtrain, Xtest, Ytrain, Ytest, dataDir, currDir):
+        '''
+        The Dense - IHT - Sparse Retrain Routine for Bonsai Training
+        '''
+        resultFile = open(dataDir + '/BonsaiResults.txt', 'a+')
+        numIters = Xtrain.shape[0] / batchSize
+
+        totalBatches = numIters * totalEpochs
+
+        counter = 0
+        if self.bonsaiObj.numClasses > 2:
+            trimlevel = 15
+        else:
+            trimlevel = 5
+        ihtDone = 0
+
+        header = '*' * 20
+        for i in range(totalEpochs):
+            print("\nEpoch Number: " + str(i))
+
+            trainAcc = 0.0
+            numIters = int(numIters)
+            for j in range(numIters):
+
+                if counter == 0:
+                    msg = " Dense Training Phase Started "
+                    print("\n%s%s%s\n" % (header, msg, header))
```

Comment:
Please accept a `files` argument to the print statement as part of your `trainer` routine. As things current stand, the only way for me to redirect console output is to redirect it using `>` .
`print` automatically accepts the `file` argument which can defaults to `None` and redirects to `sys.stdout`. Can you add an optional argument `outfilePointer` or something to your train routine or the class itself which defaults to `None`?

In case someone wants to redirect to file, he can provide a file pointer to do the same.

Reply:
I generally use the > and 2> redirection, but sure we can pass a file as an argument.

Reply:
I think this is fixed now. You can give an output file as an argument and get along with it. It defaults to sys.stdout.

Let me know if this is not fine